Refactor jupyter schema

ka-sarthak opened the pull request:
- Schema
  - [x] allow easy extension/override for writing pre-defined cells
  - [x] use tags to identify pre-defined cells
  - [x] remove input entry class
  - [x] Make `query_for_inputs` as a list. This allows to have a way of using OR operation in the search
  - [x] rename to JupyterAnalysis
- Jupyter Notebook
  - [x] Use descriptive tags for the notebooks
  - [x] entry id should be served directly to the get_analysis_entry
- [x] Move `get_entry_data` to utils
- [x] `get_entry_data` returns only the data and not everything (check if updating the entry with the data section works)
- [x] Use `method` quantity to populate the type of analysis. Remove `analysis_type`
- [x] Use `ActionEditQuantity` to generate notebook. Do not generate the notebook in the first reprocess.

ka-sarthak committed:
One notebook generation method

ka-sarthak committed:
Cleaning

ka-sarthak committed:
Fix

ka-sarthak committed:
Move get_analysis_entry to utils

ka-sarthak committed:
Use utils import

ka-sarthak committed:
Move the writing of XRD specific cells XRD analysis schema

ka-sarthak committed:
Add test for XRD

ka-sarthak committed:
Fix

ka-sarthak committed:
Redefine the notebook

ka-sarthak committed:
Xraydiffraction cells

ka-sarthak committed:
Use tags to identify the predefined cells

ka-sarthak committed:
save entry method

ka-sarthak committed:
Remove input entry class quantity

ka-sarthak committed:
Rename classes

ka-sarthak committed:
Ruff

ka-sarthak committed:
Return entry data section

ka-sarthak committed:
Use method instead of analysis_type

ka-sarthak committed:
Generate notebook action

ka-sarthak committed:
Remove reset name method

ka-sarthak committed:
Ruff

ka-sarthak committed:
Docs

ka-sarthak commented:
@hampusnasstrom Here are some changes to the Jupyter analysis schemas building upon our discussions from last week. Let me know your thoughts.

ka-sarthak committed:
Update docs, labels

ka-sarthak committed:
Allow a list of queries: mimiks OR operation

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
+                    continue
+                new_notebook.cells.append(cell)
+        else:
+            # add some empty cells
+            for _ in range(3):
+                new_notebook.cells.append(nbf.v4.new_code_cell())
```
What's the reason for adding 3 empty cells?

ka-sarthak replied:
It's meant to leave some empty cells in the end, just a design choice.

hampusnasstrom replied:
I would leave that out unless you feel strongly about it as repeatedly clicking the action button will now just add a bunch of empty cells every time.

ka-sarthak replied:
It only adds the empty cells during the first creation of the notebook. If the notebook already exists, the else clause will not run. 

I think it's nicer to have some cells in the end. Indicates the user they can continue working further on the notebook

hampusnasstrom replied:
Okay but then I would just add one. Three seems excessive.

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
-        Generates the notebook `ELNJupyterAnalysis.ipynb` and saves it in `raw` folder.
+        Generates the notebook and saves it in `raw` folder when button associated with
+        `action_trigger` is clicked. If a notebook already exists, it will only
+        overwrite the cells containing the tag `nomad-analysis-predefined`. All other
```
As it is now we are not really overwriting the predefined but copying over everything that is not predefined. How does this affect existing outputs? Maybe okay but we should then update the doc string to reflect this.

ka-sarthak replied:
The outputs of the predefined cells will be empty. The output of other cells will be retained as it is. From the user's perspective, the predefined cells be initialized again

hampusnasstrom replied:
Okay, but I would still change this docstring (that will anyways only be seen by contributors to this plugin) to reflect what is actually happening in the code. Now I read the docstring and was then quite confused by the code as it did something different.

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
+            file_name = (
+                self.name.replace(' ', '_').lower()
+                + '_'
+                + self.method.replace(' ', '_').lower()
```
Do we need to include the method in the name? I was using the name of the archive.json (without the archive.json) as this is guaranteed unique. I just foresee a lot of `_Generic.ipynb`. But not too important.

ka-sarthak replied:
Yeah, I can substitute it with `archive.metadata.mainfile`

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
-        super().normalize(archive, logger)
-
-        self.set_jupyter_notebook_name(archive, logger)
         self.normalize_input_references(
             self.process_query_for_inputs(archive, logger), logger
```
This seems a bit nested. Why not just call `process_query_for_inputs` in `normalize_input_references`?

hampusnasstrom replied:
If you do want to separate the two I would call the first function first and save the result and pass it to the second one.

ka-sarthak replied:
I will go with your first suggestion

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
+        Extends the pre-defined cells with XRD specific analysis functions.
+        """
+
+        cells = super().write_predefined_cells(archive, logger)
```
Nice!

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -293,8 +219,10 @@ def process_query_for_inputs(
         entries = []
 
         # extend the entries with the data from query_for_inputs
-        if self.query_for_inputs is not None:
-            entries.extend(self.query_for_inputs['data'])
+        if self.query_for_inputs:
+            for query in self.query_for_inputs:
+                if query.get('data') is not None:
+                    entries.extend(query['data'])
```
This looks a little bit awkward. Maybe something like:
```suggestion
        # extend the entries with the data from query_for_inputs
        if self.query_for_inputs:
            entries = [query['data'] for query in self.query_for_inputs if 'data' in query]
```

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
-            code = '# Pre-defined block\n\nxrd_voila_analysis(analysis.data.inputs)\n'
-            cells.append(nbf.v4.new_code_cell(source=code))
+        source = [
+            '# Run the cell to get the analysis entry data linked with this notebook\n',
+            'from nomad_analysis.utils import get_entry_data\n',
```
I would add this as a markdown cell and not as a comment in the python code.

hampusnasstrom replied:
The comment that is, not the import of course...

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
+            'from nomad_analysis.utils import get_entry_data\n',
+            '\n',
+            f'analysis = get_entry_data(entry_id="{archive.entry_id}")\n',
+            'analysis\n',
```
Not sure it's necessary to add a print of the analysis but you can leave it if you think it's good

hampusnasstrom commented on src/nomad_analysis/jupyter/schema.py:
```diff
@@ -410,178 +338,216 @@ def write_predefined_cells(
+            '\n',
+            'This notebook has been generated by a NOMAD Analysis entry with the\n',
+            'following definition path:\n',
+            f'`{self.m_to_dict()["m_def"]}`',
```
I get a key error when running this in the Python REPL:
```
>>> test = JupyterAnalysis()
>>> test.m_def
nomad_analysis.jupyter.schema.JupyterAnalysis:Section
>>> test.m_to_dict()["m_def"]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'm_def'
```

hampusnasstrom replied:
I think this might be a better solution:

```suggestion
            f'`{self.m_def.qualified_name()}`',
```

```
>>> test.m_def.qualified_name()
'nomad_analysis.jupyter.schema.JupyterAnalysis'
```

ka-sarthak replied:
I think `self` points to the EntryArchive which should have `m_def`

hampusnasstrom replied:
I'm pretty sure `self` should point to the section. Otherwise there would be no need to supply archive argument to the normalize method.

ka-sarthak replied:
Hmm, I am pretty sure this was working when I tested before. Good catch, and your suggestion above works as well, better than deserializing anyways

ka-sarthak replied:
I just checked, `self.m_to_dict()["m_def"]` works with the `ServerContext`. Nevertheless I like your way better.

hampusnasstrom replied:
Ah, okay. That explains it. But yes, this should be a bit more efficient than serializing the whole section.

hampusnasstrom commented on src/nomad_analysis/utils.py:
```diff
@@ -105,6 +107,48 @@ def list_to_string(list_instance: list) -> str:
+        EntryArchive: Entry archive of the analysis entry.
+    """
+
+    entry_list = []
+    try:
+        a_query = ArchiveQuery(
+            query={
+                'entry_id:any': [entry_id],
+            },
+            required='*',
+            url=url,
+            username=username,
+            password=password,
+        )
+        entry_list.extend(a_query.download(1))
```
I'm a bit confused what is happening here. Why are you initializing a list and then extending it?
```suggestion
    try:
        a_query = ArchiveQuery(
            query={
                'entry_id:any': [entry_id],
            },
            required='*',
            url=url,
            username=username,
            password=password,
        )
        entry_list = a_query.download(1)
```

hampusnasstrom replied:
Sorry, that comment sounded a bit strange. I was not necessarily confused by the extra list definition. More about how `download(1)` worked. But I guess that always returns a list but it's not always the length you ask for? Very confusing implementation....

ka-sarthak replied:
Yes, it always returns a list. The user can download a fixed number of entries at a time by specifying the length. As the entries are downloaded, they are cleared from a queue. 

With this implementation of the util, the goal was to generate different print statements when an error in ArchiveQuery is encountered and when the entry list is empty.

hampusnasstrom reviewed (approved):
Really great job @ka-sarthak :raised_hands: 
I especially like the tags, the one line getting of the analysis entry, and the extensibility of the default cells. The comments I added are just general code improvements and shouldn't change anything in implementations (except maybe the name of the file) so feel free to merge when you feel ready.

ka-sarthak committed:
Update src/nomad_analysis/jupyter/schema.py

Co-authored-by: Hampus Näsström <hampus.nasstrom@gmail.com>

ka-sarthak committed:
Review

ka-sarthak committed:
Allow resetting of inputs

ka-sarthak committed:
Add tests

ka-sarthak commented:
Thanks for the detailed review. I integrated many of the suggestions and added some tests for schema, aliasing, and schema extension. 

In addition, I added a new action button to reset the `inputs` sub-section. This will come in handy if a user mistakenly added a lot of inputs via `query_for_inputs`.
